### PR TITLE
Other file formats

### LINKED ISSUE #77: Allow VCFWriter to accept output file type like SAMWriter

### PR DESCRIPTION
**Pull request by charlesgregory**
### Single Line Reader Support
- ```RecordReader``` and ```RecordReaderRegion``` ranges
  - ```RecordReader``` handles full reading of any single line record file type
    - Record type must be provided and accept a string in its constructor
  - ```RecordReaderRegion``` handles tabix filtering of any single line record file type
    - Very similar to ```RecordReader```
- Since we are relying on ```BGZFile``` and ```TabixIndexedFile``` (really just ```hts_file```) we get automatic bgzf, gzip, https, and s3 support.
### GFF Support
- gff3d was brought into ```dhtslib``` as ```dhtslib.gff```
   - ```GFF3_Record``` was rewritten as ```GFFRecord```
     - Now can also be built from scratch and is mutable
     - Handles GFF2 and GFF3 records
 - Now have ```GFF2Reader``` and ```GFF3Reader```
   - This is an abstraction of ```RecordReader``` and ```RecordReaderRegion```
### BED Support
- ```BedRecord``` 
  - ```BedRecord``` is based off UCSC bed format
    - should be compatible with any bed file though
    - implementation very similar to ```GFFRecord```
- ```BEDReader```
     - This is an abstraction of ```RecordReader``` and ```RecordReaderRegion```
### Changes to ```VCFReader```
- ```VCFReader``` now accepts ```TabixIndexedFile``` and a ```ChromCoordinates``` region
  - This provides filtered vcf output
- This is done via ```VCFReaderImpl``` which handles the bulk of the code.
### Misc
- Reorganized ```dhtslib.vcf``` into a sub package
  - This is similar organization to the ```dhtslib.sam```, ```dhtslib.gff```, and ```dhtslib.bed``` subpackages
- github workflows has been updated to make some bgzipped and tabix'd files for unittests

Let me know what you think this will close  #77, #76, #68, #73. And gets us much closer to next release.

### CODE REVIEW

File: source/dhtslib/gff/package.d [hunk 1]
```diff
@@ -0,0 +1,176 @@
+/**
+
+GFF3 Record abstraction
+
+This module provides a readable, writeable abstraction of GFF3 records.
+
+For reading and writing from and to bgzip-compressed/tabix-indexed files,
```
Comment (jblachly) [inline source/dhtslib/gff/package.d:7]: looks like straight pull from gff3d? need to remove

**Comment by jblachly** [comment]
I see GFFRecord is mutable, but didn't see a `GFFWriter` ?

Also how does `unpack;` work?

File: source/dhtslib/gff/package.d (hunk 1 above)
Comment (charlesgregory) [reply source/dhtslib/gff/package.d:7]: Do you mean remove all gff3d code or just the descriptive text? I would have just updated gff3d and included it as a dependency. The problem is gff3d would need access to dhtslib.coordinates and that creates a cyclic dependency that dub can't solve.

**Comment by charlesgregory** [comment]
> I see GFFRecord is mutable, but didn't see a `GFFWriter` ?
> 
> Also how does `unpack;` work?

Since bed and gff are text based formats, my intention was to have no writers as you can print the header lines (lines with a #) and just print the `toString` methods of the records. Though adding simple writers (without compression support) would be fairly trivial to accomplish.

My idea with `unpack` was to keep in the functionality from `gff3d` of range-based access to the underlying string data. Though in order to have (easier) mutability, we have to unpack the data into fields and be able to overwrite that data. I thought of the idea as kind of similar to how bcf records unpack. If we assume CPUs are good at branch prediction, checking whether the data is unpacked should be pretty fast. If we are only reading data in, we use the range-based ```splitter``` to get data from the underlying `string`, however if we need to modify the data we unpack the fields into an array for easier access and modification.

**Comment by charlesgregory** [comment]
Alright I have added a `RecordWriter` struct that is pretty basic but seems to work for all of our record types. Gonna go ahead and merge this in.

**Pull request merged**